### ISSUE TITLE
Downed resolver can cause `func (s *DNS01Solver) Wait(...)` to give up too soon

### ISSUE THREAD

**Issue by cameronelliott** [caddyserver/certmagic#136]
Hey @mholt , @francislavoie, @Gansgar, I am using a copy of the dns-propagation-checker in a different codebase,
and I keep learning stuff about how it works (and when it doesn't).
I think I have found a subtle, rare, but important bug that is worth discussing how to fix.
I'm not really sure how to best interface to you guys, since I'm an unknown to you.

I'd like to get your feedback on whether you want to discuss this, understand it, maybe you just want a PR,
I'm really not sure.
Please guide me how to best help this, and let me know your receptiveness to a fix, etc.
Thanks, Cameron

##Short summary
*When the last resolver-system in the resolverslist is down, an i/o-timeout will propagate up and cause the checking-loop to abort checking*
The checking loop should continue checking until it's time limit has been reached.

##Detailed description

I think there is a possible bug in this function chain:
`func (s *DNS01Solver) Wait(...) error`
`checkDNSPropagation`
`sendDNSQuery`
`dnsQuery`

I'd like to see if there are other eyes interested in double checking what's happening, before I propose a fix, and/or a PR.

(please forgive my capitalization)
`sendDNSQuery` will try a udp dns query, and if it fails, it will try a tcp dns query.
If a tcp query fails, it will return the error code up.

`dnsQuery` is the sole caller of `sendDNSQuery`.
It will loop through the nameservers, and try the query
against each, ignoring any error returned, EXCEPT in
the case of the last nameserver.
If the last nameserver is down, then both udp and tcp queries will
fail, but the i/o-timeout error will be returned from `sendDNSQuery` to `dnsQuery`, and then to the caller of `dnsQuery`.

When the caller of `dnsQuery` is `checkDNSPropagation`, the error is again
propagated up, to `func (s *DNS01Solver) Wait(...) error`
which is doing a loop of retries 2 seconds apart.
BUT IF THE LAST RESOLVER IN THE LIST IS DOWN, THE PROPAGATED ERROR WILL CAUSE THE RETRY LOOP TO EXIT!
SHOULD ANY RESOLVERS BEING DOWN CAUSE `Wait()` TO GIVE UP? PROBABLY NOT.

**Comment by mholt** [comment]
You might be onto something here. Have you experienced this problem with waiting for DNS propagation before?

**Comment by cameronelliott** [comment]
Matt,
Have others experienced issues with DNS propagation??
I have experienced it before.


I had previously reported on an issue, where Certmagic was either slow or failing,
but due in part to bugs in the Google Wifi router with IPv6 name resolution.
The interesting thing about that issue, was I realized that with a lot of hard
work the Certmagic "dns propagation checker" could be both faster, and more
robust to environmental situations, and also a class of errors, that sometimes turn up.
https://github.com/caddyserver/certmagic/issues/122#issuecomment-822877149

Yesterday, I was seeing a failure (not using Certmagic, but using the same code from
Certmagic) that caused me to track down what I do think is a real bug, as is described in this issue.

Here is the short list of improvements I think the  "dns propagation checker" code benefit from:

- somewhat parallel checking of the various resources rather completely serial checking 
- including support for  native operating system ([cgo vs pure go DNS](https://golang.org/pkg/net/#hdr-Name_Resolution)) (**I believe, in some situations with tight firewalls, the O/S resolver may work where the by-hand pure-go resolver in Certmagic might fail**)
- For my uses I am check A/AAAA records not just TXT like Certmagic, so I am re-using the code.
- Ideally, but maybe low ROI, we would have a set of unit tests with a simple udp/tcp resolver available to test the functionality. I'm not sure the time investment will ever be worth it for me to do this, but let's talk about ideal anyway.

So, in short, I have seen the propagation-checker fail in two different situations where it could have worked if it was hardened significantly.
- #122 I believe if the dns-propagation-checker (DPC) also supported cgo/operating-system, then the DPC would have been successful instead of being super-slow (minutes). I also think a well designed semi-parallel checker may have also avoided the bugs/issues I saw in #122
- This issue. If the last resolver is down, the DPC gives up, because an error propagates all the way up to the check/sleep/check loop.  I think this is a much simpler fix than the improvements I outlined.

There is my brain dump, hope it's not too much!
Cameron

**Comment by cameronelliott** [comment]
> You might be onto something here. Have you experienced this problem with waiting for DNS propagation before?

I touched on this in my long email, but it is easy to conceive of situations with tight firewalls where the Certmagic propagation-checker is failing, or super-slow, but the system DNS seems to be working fine. This is because the Certmagic check avoids using the operating system resolver completely. Thus, there are scenarios, where the Certmagic resolver will not work, but if the cgo/operating system resolve was used from Go, thing would work fine.
I know, sorry these are so detailed, this is tricky stuff.

**Comment by mholt** [comment]
Ok, yeah, what you're saying makes sense. The DNS util code is the last remaining code borrowed from lego, which we stopped using over a year ago. I just didn't want to touch it. :grimacing: We get some complaints or reports of problems with it from time to time, and it could definitely use a revamp / rewrite. Preferably to simplify the code as much as possible, too, like if we use the system resolver. (I'm not entirely sure why we do it the way we do it.)

If you'd be willing to work on it, I'd like to review a pull request!

**Comment by cameronelliott** [comment]
Let me ask you a quick question about this.

How often do you get reports of slowness or issues with dns-name-propagation-checking?

Cameron

**Comment by francislavoie** [comment]
> How often do you get reports of slowness or issues with dns-name-propagation-checking?

From reading all the questions on https://caddy.community for the past year+, essentially never.

**Comment by cameronelliott** [comment]
I just again experience a fatal bug with Certmagic, due to external factors with DNS resolver infrastructure.

The issue is as described above, in short:
I have a resolver supporting UDP fine, but which isn't talking TCP, and this causes a fatal exit from Certmagic because the TCP dial times out.
```
panic: fooabbb.ddns5.com: obtaining certificate: [fooabbb.ddns5.com] Obtain: [fooabbb.ddns5.com] solving challenges: waiting for solver certmagic.solverWrapper to be ready: checking DNS propagation of _acme-challenge.fooabbb.ddns5.com: dial tcp [2604:a880:0:1010::12:1]:53: i/o timeout (order=https://acme-v02.api.letsencrypt.org/acme/order/127953885/10597459673) (ca=https://acme-v02.api.letsencrypt.org/directory)
```

So, basically, I think the Lego stuff works very well if your external DNS infrastructure is without any issues.

It makes me a little crazy, cause I can easily fix the TCP support question on my resolver, 
but it doesn't fix this low level library issue which should get fixed.

I'll keep chipping away at this.

**Comment by cameronelliott** [comment]
I think this can be closed, as I am not expecting you guys to hack on this, as it probably works >99% of time.
(Only when the last resolver in the list is totally down, or down on TCP will this issue occur)
I may at some point submit a PR for the dns-prop-checker, but only when I get to the point where I want to try
to squeeze a seconds or two of the delay. Which is not yet.


Thank you for listening, and building CM!
